heartbeat/service.go

```diff
@@ -0,0 +1,192 @@
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package heartbeat
+
+import (
+	"context"
+	"sync"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger/apply"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// Service emits keep-alive heartbeats for accts that are in danger of
+// suspension.
+type Service struct {
+	// addresses that should be monitored for suspension
+	accts participants
+	// current status and balances
+	ledger ledger
+	// where to send the heartbeats
+	bcast txnBroadcaster
+
+	// infrastructure
+	ctx      context.Context
+	shutdown context.CancelFunc
+	wg       sync.WaitGroup
+	log      logging.Logger
+}
+
+// NewService creates a heartbeat service. It will need to know which accounts
+// to emit heartbeats for, and how to create the heartbeats.
+func NewService(accts participants, ledger ledger, bcast txnBroadcaster, log logging.Logger) *Service {
+	return &Service{
+		accts:  accts,
+		ledger: ledger,
+		bcast:  bcast,
+		log:    log.With("Context", "heartbeat"),
+	}
+}
+
+// Start starts the goroutines for the Service.
+func (s *Service) Start() {
+	s.ctx, s.shutdown = context.WithCancel(context.Background())
+	s.wg.Add(1)
+	s.log.Info("starting heartbeat service")
+	go s.loop()
+}
+
+// Stop any goroutines associated with this worker.
+func (s *Service) Stop() {
+	s.log.Debug("heartbeat service is stopping")
+	defer s.log.Debug("heartbeat service has stopped")
+	s.shutdown()
+	s.wg.Wait()
+}
+
+// findChallenged() returns a list of accounts that need a heartbeat because
+// they have been challenged.
+func (s *Service) findChallenged(rules config.ProposerPayoutRules, current basics.Round) []account.ParticipationRecordForRound {
+	ch := apply.FindChallenge(rules, current, s.ledger, apply.ChRisky)
+	if ch.IsZero() {
+		return nil
+	}
+
+	var found []account.ParticipationRecordForRound
+	for _, pr := range s.accts.Keys(current) { // only look at accounts we have part keys for
+		acct, _, _, err := s.ledger.LookupAccount(current, pr.Account)
```

Comment:
Some of the stuff must be current, like LastHeartbeat and LastProposed, so we need a LookupAccount.

But the reason we use the current VoteID is because that's what the heartbeat transaction does, and the reason it does, is because:
```
	// Note the contrast with agreement. We require the account's _current_
	// partkey be used to sign the heartbeat. This is required because we can
	// only look 320 rounds back for voting information. If a heartbeat was
	// delayed a few rounds (even 1), we could not ask "what partkey was in
	// effect at firstValid-320?"  Using the current keys means that an account
	// that changes keys would invalidate any heartbeats it has already sent out
	// (that haven't been evaluated yet). Maybe more importantly, after going
	// offline, an account can no longer heartbeat, since it has no _current_
	// keys. Yet it is still expected to vote for 320 rounds.  Therefore,
	// challenges do not apply to accounts that are offline (even if they should
	// still be voting).
```
Is that what you mean?

Reply:
I guess I can use LookupAgreement. It lack `Status`, but if I get a response with non-zero Algos, that implies "Online", right?

You're expecting this performs better because online accounts are in cache, so I'm not hitting db?

Reply:
Done in https://github.com/algorand/go-algorand/pull/6189/commits/923a5147ad524493842f58676fc7afd19f12a380